### PR TITLE
Add the `@public` macro

### PR DESCRIPTION
**Pull request by DilumAluthge**

### CODE REVIEW

File: src/Public.jl [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+    (length(arg1_str) >= 2) || return false
+
+    # The second argument must be a `LineNumberNode`
+    (expr.args[2] isa LineNumberNode) || return false
+
+    return true
+end
+
+# export public
+# @public @public
```
Comment (goerz) [inline src/Public.jl]: `export` already marks members as public, right?

Comment (DilumAluthge) [reply src/Public.jl]: Correct, so we only need one of either `export @public` or `@public @public`.

For now, I went with `export @public`.

**Comment by DilumAluthge** [comment]
The basic functionality is now done. Let's merge this, and then I'll set up CI in a follow-up PR.

**Comment by KristofferC** [comment]
Isn't a whole package for a new feature a bit "excessive"? And if we wanted this to be a macro couldn't the original feature just have been a macro?

If you want backwards compat I would assume you would just put an if for the version and then have an eval + a `Meta.parse` like you do to support all the other new syntaxes.

**Comment by DilumAluthge** [comment]
It's kind of annoying to have to do eval+parse in package code.

Ideally, in package code, I could just do this:
```julia
@static if VERSION >= v"1.11-"
    public foo, bar
end
```

But that doesn't work.

> And if we wanted this to be a macro couldn't the original feature just have been a macro?

Actually, yeah if we could make the original feature (in Base) a macro, then I could do this in package code:

```julia
@static if VERSION >= v"1.11-"
    Base.@public foo, bar
end
```

**Comment by DilumAluthge** [comment]
I think that's my preferred solution here. In Julia 1.11, add a macro to Base of the form `Base.@public` such that `@public foo, bar` is equivalent to `public foo, bar`. And then package code just does:

```julia
@static if VERSION >= v"1.11-"
    Base.@public foo, bar
end
```

And then there's no need for this package (Public.jl), and package authors don't also need to take a dependency on Compat.jl.

**Comment by goerz** [comment]
Can that work without the `Base.` prefix in front of `@public`?

**Comment by DilumAluthge** [comment]
Sure, we can export `@public` from Base, and then in package code you'd just do:

```julia
@static if VERSION >= v"1.11-"
    @public foo, bar
end
```

**Comment by goerz** [comment]
That seems quite appealing! I’d definitely use that

**Comment by KristofferC** [comment]
> It's kind of annoying to have to do eval+parse in package code

Yeah a bit but adding a whole package (that will end up in almost every manifest for a very long time) is also kind of annoying. 

A macro in Base or adding to Compat.jl instead would be preferable in my opinion.

**Comment by fredrikekre** [comment]
This feature already exist in Compat.jl (which I thought was the JuliaLang solution to compatibility problems). Doesn't really seem sustainable to maintain a package like this per new feature.

**Comment by LilithHafner** [comment]
In 3-8 years, or whenever folks start dropping compat for 1.10, I prefer
```julia
export baz
public foo, bar
```
over
```julia
export baz
@public foo, bar
```

In the interem, I think
```julia
using Compat
export baz
@compat public foo, bar
```
is fine; the point of Compat is to allow Base to add new features and let folks get those features while keeping compat from before they were implemented, and I think it does a good enough job.

Today, there may be a large number of packages which _only_ want Compat.jl for the public keyword, so it is tempting to excise that functionality into another package and write

```julia
using Public
export baz
@public foo, bar
```

But as time passes and the public keyword stops being the most recent major Compat.jl addition, it will be increasingly common for folks who need `@public` to also need other compat features.

FWIW, on my computer, Compat.jl takes 8 seconds to install as a standalone package (mostly downloading the general registry), 42ms to load the first time, and, after restarting Julia, 3.6ms to load. It is also already in most large manifests.

> Doesn't really seem sustainable to maintain a package like this per new feature.

In theory, a micro package could have a near-zero maintenance burden. If would be nice if that were actually the case.

**Comment by KristofferC** [comment]
Is writing `eval(Meta.parse` really that much worse than adding a whole dependency to your package...? 

Anyway, whatever solution is good that doesn't cause this type of one package per feature thing to happen. If that is a macro in Base or in Compat or whatever doesn't matter much to me.

**Comment by goerz** [comment]
> Is writing eval(Meta.parse really that much worse

How does that look, exactly? Something like this?

```julia
@static if VERSION >= v"1.11-"
eval(Meta.parse, quote
    public foo, bar
end)
end
```

**Comment by KristofferC** [comment]
`eval(Meta.parse("..."))`

For example https://github.com/JuliaLang/Tokenize.jl/blob/13396b85c2847544e9231cebb509de5c5674e46b/src/token_kinds.jl#L1497

**Comment by LilithHafner** [comment]
You can also do
```julia
VERSION >= v"1.11.0-DEV.469" && eval(Expr(:public,
    :foo,
    :bar
))
```

if you want to drop that pesky dependency on `Meta.parse`. :smile: 

https://github.com/JuliaLang/julia/blob/bdd3ffdd215db8233a43b6ba34c709add2e26b3c/base/exports.jl#L1068-L1073

Edit to clarify: unless you are doing some extreme tree-shaking for binary size, or have some other reason to prefer this approach, it's not a great idea. "...that pesky dependency on `Meta.parse`" is a joke, though in bootstrapping can be less of a joke.

**Comment by KristofferC** [comment]
Seems worse to rely on some (undocumented?) Expr `head`..

**Pull request merged**